adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -1082,28 +1083,23 @@ public final void uploadCommandResponseMessage(
 
         final CommandResponse commandResponse = CommandResponse.from(commandRequestId, tenant, deviceId, payload,
                 contentType, responseStatus);
-        if (commandResponse == null) {
-            metrics.reportCommand(
-                    Direction.RESPONSE,
-                    tenant,
-                    ProcessingOutcome.UNPROCESSABLE,
-                    payload.length(),
-                    getMicrometerSample(ctx));
-            final String errorMsg = String.format("command-request-id [%s] or status code [%s] is missing/invalid",
-                    commandRequestId, responseStatus);
-            TracingHelper.logError(currentSpan, errorMsg);
-            currentSpan.finish();
-            HttpUtils.badRequest(ctx, errorMsg);
-        } else {
+
+        final Future<CommandResponse> commandResponseFuture = commandResponse != null
+                ? Future.succeededFuture(commandResponse)
+                : Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                        String.format("command-request-id [%s] or status code [%s] is missing/invalid",
+                                commandRequestId, responseStatus)));
+
+        final Future<ProtonDelivery> uploadMessageTracker = commandResponseFuture.compose(resp -> {
             final Future<JsonObject> deviceRegistrationTracker = getRegistrationAssertion(
                     tenant,
                     deviceId,
                     authenticatedDevice,
                     currentSpan.context());
             final Future<Void> tenantEnabledTracker = getTenantConfiguration(tenant, currentSpan.context())
                     .compose(tenantObject -> isAdapterEnabled(tenantObject).map(ok -> null));
-            CompositeFuture.all(deviceRegistrationTracker, tenantEnabledTracker)
-                    .compose(ok -> sendCommandResponse(tenant, commandResponse, currentSpan.context()))
+            return CompositeFuture.all(deviceRegistrationTracker, tenantEnabledTracker)
+                    .compose(ok -> sendCommandResponse(tenant, commandResponseFuture.result(), currentSpan.context()))
```

Comment:
I don't understand the reason to use the future here. I get the reason for that in general, but in this location I think you could/should use `resp` instead, which should be the same value, unless I missed something.

If I understood it correctly, then technically it should't make a difference if you use `resp`, `commandResponseFuture.result()` or `commandResponse` here. However code might change over time, and this would seem more robust to me.

Reply:
Ok, I've changed it to `resp` and then renamed it to `commandResponse` (renaming the other `commandResponse`).